### PR TITLE
Add ConvEmformer module

### PR DESCRIPTION
**Pull request by hwangjeff**
Adds an implementation of the convolution-augmented streaming transformer (effectively Emformer with convolution block) described in https://arxiv.org/abs/2110.05241.

Continuation of #2324.

### CODE REVIEW

**Comment by hwangjeff** [comment]
Docs: https://output.circle-artifacts.com/output/job/718cda5f-f5bf-400d-89a1-66aaad6727e6/artifacts/0/docs/prototype.models.html#convemformer

File: torchaudio/prototype/models/conv_emformer.py [hunk 1]
```diff
@@ -431,3 +438,87 @@ def infer(
...
+        tanh_on_mem (bool, optional): if ``True``, applies tanh to memory elements. (Default: ``False``)
+        negative_inf (float, optional): value to use for negative infinity in attention weights. (Default: -1e8)
+        conv_activation (str, optional): activation function to use in convolution modules.
+            Must be one of ("relu", "gelu", "silu"). (Default: "silu")
+
+    Examples:
+        >>> conv_emformer = ConvEmformer(80, 4, 1024, 12, 16, 8, right_context_length=4)
+        >>> input = torch.rand(10, 200, 80)
+        >>> lengths = torch.randint(1, 200, (10,))
+        >>> output = conv_emformer(input, lengths)
```
Comment (nateanl) [inline torchaudio/prototype/models/conv_emformer.py:472]:
```suggestion
        >>> output, lengths = conv_emformer(input, lengths)
```

File: torchaudio/models/emformer.py [hunk 2]
```diff
@@ -857,3 +798,79 @@ def infer(
...
+        ffn_dim (int): hidden layer dimension of each Emformer layer's feedforward network.
+        num_layers (int): number of Emformer layers to instantiate.
+        segment_length (int): length of each input segment.
+        dropout (float, optional): dropout probability. (Default: 0.0)
+        activation (str, optional): activation function to use in each Emformer layer's
+            feedforward network. Must be one of ("relu", "gelu", "silu"). (Default: "relu")
+        left_context_length (int, optional): length of left context. (Default: 0)
+        right_context_length (int, optional): length of right context. (Default: 0)
+        max_memory_size (int, optional): maximum number of memory elements to use. (Default: 0)
+        weight_init_scale_strategy (str, optional): per-layer weight initialization scaling
```
Comment (nateanl) [inline torchaudio/models/emformer.py:820]:
```suggestion
        weight_init_scale_strategy (str or None, optional): per-layer weight initialization scaling
```

File: torchaudio/models/emformer.py [hunk 3]
```diff
@@ -857,3 +798,79 @@ def infer(
...
+        num_heads: int,
+        ffn_dim: int,
+        num_layers: int,
+        segment_length: int,
+        dropout: float = 0.0,
+        activation: str = "relu",
+        left_context_length: int = 0,
+        right_context_length: int = 0,
+        max_memory_size: int = 0,
+        weight_init_scale_strategy: str = "depthwise",
```
Comment (nateanl) [inline torchaudio/models/emformer.py:847]:
```suggestion
        weight_init_scale_strategy: Optional[str] = "depthwise",
```

File: torchaudio/models/emformer.py [hunk 4]
```diff
@@ -857,3 +798,79 @@ def infer(
...
+        weight_init_scale_strategy (str, optional): per-layer weight initialization scaling
+            strategy. Must be one of ("depthwise", "constant", ``None``). (Default: "depthwise")
+        tanh_on_mem (bool, optional): if ``True``, applies tanh to memory elements. (Default: ``False``)
+        negative_inf (float, optional): value to use for negative infinity in attention weights. (Default: -1e8)
+
+    Examples:
+        >>> emformer = Emformer(512, 8, 2048, 20, 4, right_context_length=1)
+        >>> input = torch.rand(128, 400, 512)  # batch, num_frames, feature_dim
+        >>> lengths = torch.randint(1, 200, (128,))  # batch
+        >>> output = emformer(input, lengths)
```
Comment (nateanl) [inline torchaudio/models/emformer.py:829]:
```suggestion
        >>> output, lengths = emformer(input, lengths)
```

File: torchaudio/prototype/models/conv_emformer.py [hunk 5]
```diff
@@ -431,3 +438,87 @@ def infer(
...
+        num_layers (int): number of ConvEmformer layers to instantiate.
+        segment_length (int): length of each input segment.
+        kernel_size (int): size of kernel to use in convolution modules.
+        dropout (float, optional): dropout probability. (Default: 0.0)
+        ffn_activation (str, optional): activation function to use in feedforward networks.
+            Must be one of ("relu", "gelu", "silu"). (Default: "relu")
+        left_context_length (int, optional): length of left context. (Default: 0)
+        right_context_length (int, optional): length of right context. (Default: 0)
+        max_memory_size (int, optional): maximum number of memory elements to use. (Default: 0)
+        weight_init_scale_strategy (str, optional): per-layer weight initialization scaling
```
Comment (nateanl) [inline torchaudio/prototype/models/conv_emformer.py:461]:
```suggestion
        weight_init_scale_strategy (str or None, optional): per-layer weight initialization scaling
```

File: torchaudio/prototype/models/conv_emformer.py [hunk 6]
```diff
@@ -431,3 +438,87 @@ def infer(
...
+        ffn_dim: int,
+        num_layers: int,
+        segment_length: int,
+        kernel_size: int,
+        dropout: float = 0.0,
+        ffn_activation: str = "relu",
+        left_context_length: int = 0,
+        right_context_length: int = 0,
+        max_memory_size: int = 0,
+        weight_init_scale_strategy: str = "depthwise",
```
Comment (nateanl) [inline torchaudio/prototype/models/conv_emformer.py:491]:
```suggestion
        weight_init_scale_strategy: Optional[str] = "depthwise",
```

File: torchaudio/models/emformer.py (hunk 2 above)
Comment (hwangjeff) [reply torchaudio/models/emformer.py]: nice, thanks for the catch

**Review by mthrok — APPROVED** [review]

**Review by nateanl — APPROVED** [review]

**Pull request closed without merging**